README.md
# srisum(1) -- compute and check subresource integrity digests

## SYNOPSIS

`$ srisum [OPTION]... [FILE]...`

## EXAMPLES

### Computing SRI Digests

For a single file:

```
$ srisum styles.css > styles.css.sri
```

For multiple different files:

```
$ srisum styles.css index.js package.json bundle.js > app.sri
```

From `stdin`:

```
$ cat styles.css | srisum -a sha1
sha1-hmkHOZdrfLUVOqpAgryfC8XNGtE -
```

Specify algorithms to generate:

```
$ srisum styles.css index.js --algorithms sha512 sha256 sha1 > styles.css.sri
```

### Checking Integrity

Passing checksum file as an argument:

```
$ srisum -c styles.css.sri
styles.css: OK (sha512)
```

Passing multiple checksum files:

```
$ srisum -c styles.css.sri js-files.sri
styles.css: OK (sha512)
index.js: OK (sha512)
lib/util.js: OK (sha512)
```

Checksum file from `stdin`:

```
$ cat styles.css.sri | srisum -c
styles.css: OK (sha512)
```

Checksum `stdin` itself:

```
$ echo "hello" | srisum > stdin.sri
$ echo "hello" | srisum -c stdin.sri
-: OK (sha512)
```

## DESCRIPTION

Print or check Subresource Integrity digests.

Spec: https://w3c.github.io/webappsec/specs/subresourceintegrity/

`srisum`'s API is based on the `SHA[N]SUM(1)` family of unix utilities.

With no `FILE` or when `FILE` is `-`, read standard input.

`-a, --algorithms [ALGO]...` - hash algorithms to generate for the `FILE`s

`-c, --check` - read SRI sums from the `FILE`s and check them

`-d, --digest-only` - only output the digest for each `FILE`, without filenames

`--help` - display help and exit

`--version` - output version information and exit

## The following options are useful only when verifying integrity:

`--ignore-missing` - don't fail or report status for missing files

`--quiet` - don't print OK for each successfully verified file

`--status` - don't output anything, status code shows success

`-w, --warn` - warn about improperly formatted SRI lines

When checking, the input should be a former output of this program. The default mode is to print line with space-separated SRI digests, one more space, and a name for each FILE.

Strict mode, enabled with `--strict`, will entirely ignore digests (in input and output) that fail all of the following conditions:

- `algorithms` must be one or more of: `sha1`, `sha256`, `sha384`, `sha512`
- digest strings must be valid `RFC4648` `Base64` strings.

## AUTHOR

Written by [Kat Marchán (@zkat)](https://github.com/zkat)

## CONTRIBUTING

The srisum team enthusiastically welcomes contributions and project
participation! There's a bunch of things you can do if you want to contribute!
The [Contributor Guide](CONTRIBUTING.md) has all the information you need for
everything from reporting bugs to contributing entire new features. Please
don't hesitate to jump in if you'd like to, or even ask us questions if
something isn't clear.

All participants and maintainers in this project are expected to follow [Code
of Conduct](CODE_OF_CONDUCT.md), and just generally be excellent to each
other.

Happy hacking!

## LICENSE

This project is licensed under [Apache-2.0](LICENSE-APACHE.md).


CHANGELOG.md
# `kdl` Release Changelog

<a name="5.0.0"></a>
## 5.0.0 (2023-02-03)

### Features

* **usize:** switch to usize for potential 32-bit support ([65c34e12](https://github.com/kdl-org/kdl-rs/commit/65c34e1221ab210443dd99e3d76e0d69d0b599d3))
* **deps:** bump deps ([75816120](https://github.com/kdl-org/kdl-rs/commit/75816120e7250581e136af4676188558ef25305f))

### Bug Fixes

* **clippy:** cargo clippy --fix ([13f576fd](https://github.com/kdl-org/kdl-rs/commit/13f576fdd774e37714a77e8c3a793fb97d93006e))
* **compute:** make compute output always be utf8 ([3a5e840f](https://github.com/kdl-org/kdl-rs/commit/3a5e840f9a2bf0917c422e24621358a9fb8d1589))
    * **BREAKING CHANGE**: input filenames that are not valid utf8 will not be
validly output anymore, not that they ever really were.

<a name="4.0.0"></a>

## 4.0.0 (2022-06-11)

### Features

* **modernify:** update srisum to get with the times ([256b3f9e](https://github.com/kdl-org/kdl-rs/commit/256b3f9eb88b21f44396bd7ff6b6cc15d28d109a))

<a name="3.0.0"></a>

## 3.0.0 (2019-11-06)

#### Breaking Changes

- **errors:** rewrite API to use anyhow::Result instead of panicking ([793bd9e7](https://github.com/zkat/srisum-rs/commit/793bd9e75f089f6c4a75fa4f3b5e108fa17a8487))
- **license:** upgrade to Parity 7.0 release ([850f9266](https://github.com/zkat/srisum-rs/commit/850f926686a7869eecf456bc921d725f6db96640))

#### Bug Fixes

- **compute:** properly support non-UTF8 filenames for output. Note that `check` still can't handle reading funky filenames from a checksum file/stream ([d3759096](https://github.com/zkat/srisum-rs/commit/d375909685ae5d100dc5f832b74e09cdf16c3512))

#### Features

- **errors:** rewrite API to use anyhow::Result instead of panicking ([793bd9e7](https://github.com/zkat/srisum-rs/commit/793bd9e75f089f6c4a75fa4f3b5e108fa17a8487))
- **license:** upgrade to Parity 7.0 release ([850f9266](https://github.com/zkat/srisum-rs/commit/850f926686a7869eecf456bc921d725f6db96640))

<a name="2.0.0"></a>

## 2.0.0 (2019-10-21)

#### Breaking Changes

- **license:** switch license to Parity 7.0.0-pre.3 + Apache-2.0 ([db678d74](https://github.com/zkat/srisum-rs/commit/db678d740ed61fc99082762d532c787dc1243110))

#### Bug Fixes

- **compute:** use path.display() for cross-platform printing ([823c1cd5](https://github.com/zkat/srisum-rs/commit/823c1cd5235f99294ec3d4df8bbb2b6eda486def))
- **windows:** encode_wide does u16, not u8 ([a6ab4097](https://github.com/zkat/srisum-rs/commit/a6ab40978a7a7a4fa1eae3524f9bb0974193e5fb))

#### Features

- **license:** switch license to Parity 7.0.0-pre.3 + Apache-2.0 ([db678d74](https://github.com/zkat/srisum-rs/commit/db678d740ed61fc99082762d532c787dc1243110))


src/lib.rs
mod check;
mod compute;
mod errors;
use std::ffi::OsString;

pub use check::check;
pub use compute::compute;

use clap::Parser;

#[derive(Debug, Parser)]
#[clap(author, version, about)]
pub struct CliArgs {
    /// Files to process, depending on mode.
    #[clap(name = "FILE", default_value = "-")]
    pub files: Vec<OsString>,

    /// Hash algorithms to generate for the FILEs.
    #[clap(short, long, default_value = "sha256")]
    pub algorithms: Vec<String>,

    /// Print digests only, without the filename.
    #[clap(short, long)]
    pub digest_only: bool,

    /// Read integrity checksums from the FILEs and check them.
    #[clap(short, long)]
    pub check: bool,

    /// Don't fail or report status for missing files.
    #[clap(long)]
    pub ignore_missing: bool,

    /// Don't print OK for each successfully verified file.
    #[clap(short, long)]
    pub quiet: bool,

    /// Don't output anything. Status code shows success.
    #[clap(long)]
    pub status: bool,

    /// Warn about improperly formatted checksum lines.
    #[clap(short, long)]
    warn: bool,
}


src/check.rs
use std::ffi::OsStr;
use std::fs::File;
use std::io::{self, prelude::*, BufReader, Read};

use miette::{IntoDiagnostic, Report, Result, WrapErr};
use ssri::{Algorithm, Integrity, IntegrityChecker};

use crate::errors::SrisumError;
use crate::CliArgs;

struct Stats {
    bad_lines: usize,
    bad_checksums: usize,
    missing_files: usize,
}

impl Stats {
    fn new() -> Self {
        Self {
            bad_lines: 0,
            bad_checksums: 0,
            missing_files: 0,
        }
    }
}

pub fn check(args: CliArgs) -> Result<()> {
    let files = &args.files;
    let mut stats = Stats::new();
    let marker = OsStr::new("-");
    for f in files {
        let handle: Box<dyn Read> = if marker == f {
            Box::new(std::io::stdin())
        } else {
            Box::new(File::open(f).into_diagnostic().wrap_err_with(|| {
                format!("Failed to open checksum file: {}", f.to_string_lossy())
            })?)
        };
        handle_stream(&mut stats, BufReader::new(handle), &args).context("CRITICAL ERROR")?;
    }
    print_warnings(&stats, &args);
    if stats.bad_lines > 0 || stats.bad_checksums > 0 || stats.missing_files > 0 {
        Err(miette::miette!("Checksum failed"))
    } else {
        Ok(())
    }
}

fn handle_stream<T: Read>(stats: &mut Stats, s: BufReader<T>, args: &CliArgs) -> Result<()> {
    // TODO - use s.split() to support OsStr filenames in the RHS.
    for line in s.lines() {
        // Lines must unfortunately be valid UTF-8 right now (a restriction
        // that only applies to check, but not to compute). This will be the
        // case until such a time when it's deemed necessary to go through the
        // pain of manually reading out these lines in their OsStr form. Fuck
        // that, honestly. So, we treat bad UTF-8 lines (or any other encoding error) as bad lines and just keep going.
        let line = if let Ok(line) = line {
            line
        } else {
            stats.bad_lines += 1;
            continue;
        };
        let line = line.trim();
        // Empty lines are fine, just continue.
        if line.is_empty() {
            continue;
        }
        // Split each line into `<hash>\s+<filename>`
        let split = String::from(line).chars().position(|x| x.is_whitespace());
        // No \s+ in the middle means we got a bad line. Just skip it.
        let idx = if let Some(split) = split {
            split
        } else {
            stats.bad_lines += 1;
            continue;
        };
        let (hash, filename) = line.split_at(idx);
        let filename = filename.trim();
        let sri = hash.parse::<Integrity>();
        // If the integrity string fails to parse... that's a bad line. Skip.
        let sri = if let Ok(sri) = sri {
            sri
        } else {
            stats.bad_lines += 1;
            continue;
        };
        let result = check_file(filename, sri);
        match result {
            Ok((algo, f)) => {
                if !args.status && !args.quiet {
                    println!("{f}: OK ({algo})")
                }
            }
            Err(cause) => {
                print_messages(cause, stats, args, filename)?;
            }
        };
    }
    Ok(())
}

fn print_messages(cause: Report, stats: &mut Stats, args: &CliArgs, filename: &str) -> Result<()> {
    match cause.downcast::<SrisumError>() {
        // ENOENT
        Ok(SrisumError::Io(ref e)) if e.kind() == io::ErrorKind::NotFound => {
            stats.missing_files += 1;
            if !args.ignore_missing {
                let current_exe = std::env::current_exe().into_diagnostic()?;
                let current_exe_display = current_exe.display();
                eprintln!("{current_exe_display}: {filename}: No such file or directory");
                println!("{filename}: FAILED open or read");
            }
        }
        // EINTEGRITY
        Ok(SrisumError::IntegrityError(_)) => {
            stats.bad_checksums += 1;
            if !args.status && !args.quiet {
                println!("{filename}: FAILED");
            }
        }
        Ok(err) => {
            return Err(err.into());
        }
        // Other errors
        Err(err) => {
            return Err(err);
        }
    }
    Ok(())
}

fn check_file(f: &str, sri: Integrity) -> Result<(Algorithm, String)> {
    let stream: Box<dyn Read> = if f == "-" {
        Box::new(std::io::stdin())
    } else {
        Box::new(File::open(f).into_diagnostic()?)
    };
    let mut reader = BufReader::new(stream);
    let mut checker = IntegrityChecker::new(sri);
    let mut buf = [0; 1024 * 256];
    loop {
        let amt = reader.read(&mut buf).into_diagnostic()?;
        if amt == 0 {
            break;
        } else {
            checker.input(&buf[0..amt]);
        }
    }
    let algo = checker.result().map_err(SrisumError::IntegrityError)?;
    Ok((algo, String::from(f)))
}

fn print_warnings(stats: &Stats, args: &CliArgs) {
    if args.status {
        return;
    }
    if args.warn && stats.bad_lines > 0 {
        eprintln!(
            "srisum: WARNING: {} file(s) could not be read",
            stats.bad_lines
        );
    }
    if !args.ignore_missing && stats.missing_files > 0 {
        eprintln!(
            "srisum: WARNING: {} listed file(s) could not be read",
            stats.missing_files
        );
    }
    if stats.bad_checksums > 0 {
        eprintln!(
            "srisum: WARNING: {} computed checksum(s) did NOT match",
            stats.bad_checksums
        )
    }
}


src/compute.rs
use std::ffi::OsStr;
use std::fs::File;
use std::io::{BufReader, Read};
use std::path::Path;

use miette::{IntoDiagnostic, Result, WrapErr};
use ssri::{Algorithm, Integrity, IntegrityOpts};

use crate::CliArgs;

pub fn compute(args: CliArgs) -> Result<()> {
    let files = &args.files;
    for f in files {
        let sri = hash_file(f, &args)
            .with_context(|| format!("Failed to hash file: {}", f.to_string_lossy()))?;
        if args.digest_only {
            println!("{sri}");
        } else {
            let path: &Path = f.as_ref();
            println!("{sri} {}", path.display());
        }
    }
    Ok(())
}

fn hash_file(f: &OsStr, args: &CliArgs) -> Result<Integrity> {
    let mut builder = IntegrityOpts::new();
    for algo in &args.algorithms {
        let algo = match algo.as_str() {
            "sha1" => Algorithm::Sha1,
            "sha256" => Algorithm::Sha256,
            "sha384" => Algorithm::Sha384,
            "sha512" => Algorithm::Sha512,
            _ => panic!("bad algorithm: {algo}"),
        };
        builder = builder.algorithm(algo);
    }

    if f == OsStr::new("-") {
        read_from_file(BufReader::new(std::io::stdin()), &mut builder)?;
    } else {
        read_from_file(
            BufReader::new(File::open(f).into_diagnostic()?),
            &mut builder,
        )?;
    };
    Ok(builder.result())
}

fn read_from_file<T: Read>(mut reader: BufReader<T>, builder: &mut IntegrityOpts) -> Result<()> {
    let mut buf = [0; 1024 * 256];
    loop {
        let amt = reader.read(&mut buf).into_diagnostic()?;
        if amt == 0 {
            return Ok(());
        } else {
            builder.input(&buf[0..amt]);
        }
    }
}
